File: packages/eyes-universal-poc/src/refer.js
const uuid = require('uuid')

const REF_ID = 'applitools-ref-id'

function makeRefer() {
  const store = new Map()
  const relation = new Map()

  return {isRef, ref, deref, destroy}

  function isRef(ref) {
    return Boolean(ref[REF_ID])
  }

  function ref(value, parentRef) {
    const ref = uuid.v4()
    store.set(ref, value)
    if (parentRef) {
      let childRefs = relation.get(parentRef[REF_ID])
      if (!childRefs) {
        childRefs = new Set()
        relation.set(parentRef[REF_ID], childRefs)
      }
      childRefs.add(ref)
    }
    return {[REF_ID]: ref}
  }

  function deref(ref) {
    if (isRef(ref)) {
      return store.get(ref[REF_ID])
    } else {
      return ref
    }
  }

  function destroy(ref) {
    if (!isRef(ref)) return
    const childRefs = relation.get(ref[REF_ID])
    childRefs.forEach(childRef => destroy(childRef))
    store.delete(ref[REF_ID])
  }
}

module.exports = makeRefer

File: rb/eyes-selenium/lib/applitools/refer.rb
require('securerandom')

module Applitools
  class Refer
    attr_reader :store, :relation
    REF_ID = 'applitools-ref-id'.freeze

    def initialize()
      @store = {}
      @relation = {}
    end

    def ref(value, parentRef = nil)
      uuid = SecureRandom.uuid
      store[uuid] = value
      if (parentRef)
        childRefs = relation[parentRef[:REF_ID]]
        if (!childRefs)
          childRefs = []
          childRefs.push(uuid)
          relation[parentRef[REF_ID]] = childRefs
        else
          childRefs.push(uuid)
        end
      end
      {REF_ID => uuid}
    end

    def isRef(ref)
      !!destructure_ref(ref)
    end

    def deref(ref)
      isRef(ref) ? store[destructure_ref(ref)] : ref
    end
    
    def destroy(ref)
      return if (!isRef(ref))
      childRefs = relation[destructure_ref(ref)]
      childRefs.each{|childRef| destroy({"#{REF_ID}": childRef})} if childRefs
      store.delete(destructure_ref(ref))
    end

    private

      def destructure_ref(ref)
        ref.keys.first.is_a?(Symbol) ? ref[:"#{REF_ID}"] : ref[REF_ID]
      end
  end
end
